### PR TITLE
Implement coloring for disabled branches in the shader editor

### PR DESCRIPTION
**Pull request by Chaosus**
This PR implement dimming of the disabled branches of the shader preprocessor. To do it, besides adding a handling of the branch to the shader preprocessor, I've implemented some new methods to the `CodeHighlighter`:

image

This is a part of https://github.com/godotengine/godot-proposals/issues/5062

### CODE REVIEW

**Comment by akien-mga** [comment]
Looks nice!

How come `#else` has a different color than `#ifdef` and `#endif`?

**Comment by Chaosus** [comment]
> How come #else has a different color than #ifdef and #endif?

That's strange, I will look what caused this a bit later this day.

**Comment by Chaosus** [comment]
@akien-mga It's not related or introduced by this PR, it's due preprocessor directive and keywords have similar names (#if, #else and `if`, `else`) and the control flow keywords have a separate `control_flow_keyword_color`. Not sure how it should be fixed, and I think it's a subject for other PR.

**Comment by bitsawer** [comment]
That was fast :) I guess you were already working on that before I made that improvement proposal! Looks good!

Interestingly, I think the structure ShaderPreprocessor::SkippedCondition and all logic related to that is actually dead code and mostly unused. I think TheOrangeDay was working on that to implement this very syntax highlighting feature but never finished it and now it doesn't really do anything useful. Removing it up might be a good idea, unless I am wrong. It has been a while since I touched that code, so no guarantees. Not really related to this PR either.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
`SyntaxHighlighters` are designed to be self contained rather then having "overrides" passed in or should make calls out e.g `GDScriptSyntaxHighlighter`.

Not entirely sure what that looks like, some initial thoughts, are something along the lines of the way regions work, or create a sub class e.g `GDShaderSyntaxHighlighter` for it.

**Comment by Chaosus** [comment]
@Paulb23 Done, I think - pushed new methods to a new class `GDShaderSyntaxHighlighter`.

File: scene/resources/syntax_highlighter.cpp [hunk 1]
```diff
@@ -652,3 +653,120 @@ void CodeHighlighter::set_member_variable_color(Color p_color) {
...
+	const int line_length = str.length();
+
+	for (const BranchRegion &region : disabled_branch_regions) {
+		if (p_line >= region.from_line && p_line <= region.to_line) {
+			Dictionary highlighter_info;
+			highlighter_info["color"] = disabled_branch_color;
+
+			for (int i = 0; i < line_length; i++) {
+				color_map[i] = highlighter_info;
+			}
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.cpp:672]: There shouldn't be a need to loop here as the highlighter will only stop using the current colour when it reaches a new line, or a new entry in the `color_map`. Given we're setting the entire line the same colour, `color_map[0] = highlighter_info;` should work.

File: scene/resources/syntax_highlighter.h [hunk 2]
```diff
@@ -142,4 +143,36 @@ class CodeHighlighter : public SyntaxHighlighter {
...
+///////////////////////////////////////////////////////////////////////////////
+
+class GDShaderSyntaxHighlighter : public CodeHighlighter {
+	GDCLASS(GDShaderSyntaxHighlighter, CodeHighlighter)
+
+private:
+	struct BranchRegion {
+		int from_line = -1;
+		int to_line = -1;
+	};
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.h:155]: Can use a `Point2i` rather then adding struct, unless this is for some future changes?

**Review by Paulb23 — COMMENTED** [review]
Couple notes inline, would also suggest making this editor only similar to the `GDScriptSyntaxHighlighter`.

**Comment by Chaosus** [comment]
@Paulb23 Fixed as you said.

File: scene/resources/syntax_highlighter.cpp [hunk 3]
```diff
@@ -611,6 +611,7 @@ void CodeHighlighter::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "symbol_color"), "set_symbol_color", "get_symbol_color");
 	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "function_color"), "set_function_color", "get_function_color");
 	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "member_variable_color"), "set_member_variable_color", "get_member_variable_color");
+	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "disabled_branch_color"), "set_disabled_branch_color", "get_disabled_branch_color");
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.cpp:614]: This is no longer needs to be bound.

File: scene/register_scene_types.cpp [hunk 4]
```diff
@@ -401,6 +401,7 @@ void register_scene_types() {
 	GDREGISTER_CLASS(CodeEdit);
 	GDREGISTER_CLASS(SyntaxHighlighter);
 	GDREGISTER_CLASS(CodeHighlighter);
+	GDREGISTER_CLASS(GDShaderSyntaxHighlighter);
```
Comment (Paulb23) [inline scene/register_scene_types.cpp:404]: As before no need to expose it, should be fine to have this as editor only.

File: scene/resources/syntax_highlighter.h [hunk 5]
```diff
@@ -142,4 +143,31 @@ class CodeHighlighter : public SyntaxHighlighter {
...
+	Vector<Point2i> disabled_branch_regions;
+	Color disabled_branch_color;
+
+protected:
+	static void _bind_methods();
+
+public:
+	virtual Dictionary _get_line_syntax_highlighting_impl(int p_line) override;
+
+	void add_disabled_branch_region(int p_from_line, int p_to_line);
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.h:161]: Can have this take a `Point2i` now to save us from parsing it.

File: scene/resources/syntax_highlighter.h [hunk 6]
```diff
@@ -142,4 +143,31 @@ class CodeHighlighter : public SyntaxHighlighter {
...
+	void remove_disabled_branch_region(int p_from_line);
+	bool has_disabled_branch_region(int p_from_line) const;
+
+	void set_disabled_branch_regions(const Vector<String> &p_regions);
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.h:165]: Ditto here, or as it editor only can remove these methods if we have no plans to use them.

File: scene/resources/syntax_highlighter.h [hunk 7]
```diff
@@ -142,4 +143,31 @@ class CodeHighlighter : public SyntaxHighlighter {
...
+	void clear_disabled_branch_regions();
+	Vector<String> get_disabled_branch_regions() const;
```
Comment (Paulb23) [inline scene/resources/syntax_highlighter.h:167]: `Point2i` here also.

**Review by Paulb23 — CHANGES_REQUESTED** [review]

**Comment by Chaosus** [comment]
@Paulb23 OK, I've cut all unused methods and made this class unexposed to the editor.

**Comment by Paulb23** [comment]
@Chaosus would move the highlighter code to the top of `shader_editor_plugin.[cpp|h]` as it's not relevant for scene now. Same way `EditorSyntaxHighlighter` is at the top of `script_editor_plugin`.

**Comment by Chaosus** [comment]
@Paulb23 Done

**Review by Paulb23 — APPROVED** [review]
Highlighter changes looks good.

On `How come #else has a different color than #ifdef and #endif?`

Keywords cannot contain symbols apart from `_` so other then adding an exception for the shader editor, we could try using regions something like `add_color_region('#', ' ', colour, false)`

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**